File: 028treeValueCount.js
const a = new Node(12);
const b = new Node(6);
const c = new Node(6);
const d = new Node(4);
const e = new Node(6);
const f = new Node(12);

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;

//      12
//    /   \
//   6     6
//  / \     \
// 4   6     12

// class Node {
//   constructor(val) {
//     this.val = val;
//     this.left = null;
//     this.right = null;
//   }
// }

const treeValueCount = (root, target) => {
  // todo
  if (root === null) return 0;
  let occur = 0
  const stack = [root];
  while (stack.length > 0) {
    const current = stack.shift();
    if (current.val === target) occur++;

    if (current.left) stack.push(current.left);
    if (current.right) stack.push(current.right);
  }
  return occur;
};

treeValueCount(a, 6); // -> 3

File: 028tree_value_count.py
# class Node:
#   def __init__(self, val):
#     self.val = val
#     self.left = None
#     self.right = None

# iterative solution
from collections import deque
# def tree_value_count(root, target):
#   if root is None:
#     return 0

#   queue = deque([root])
#   count = 0


#   while queue:
#     current = queue.popleft()

#     if current.val == target:
#       count += 1

#     if current.right is not None:
#       queue.append(current.right)

#     if current.left is not None:
#       queue.append(current.left)

#   return count

# recursive solution
def tree_value_count(root, target):
  if root is None:
    return 0

  match = 1 if root.val == target else 0
  return match + tree_value_count(root.left, target) + tree_value_count(root.right, target)

